torch_xla/csrc/runtime/ifrt_computation_client.cc

```diff
@@ -0,0 +1,678 @@
+  std::partial_sort_copy(client_->devices().begin(), client_->devices().end(),
+                         ordered_devices.begin(), ordered_devices.end(),
+                         [](auto& a, auto& b) { return a->id() < b->id(); });
+  for (auto* device : ordered_devices) {
+    global_ordinals_[device->id()] = global_ordinals_.size();
+    std::string device_str = IfrtDeviceToString(device);
+    string_to_device_.emplace(device_str, device);
+  }
+  comp_env_hash_ = hash_comp_env(client_, ordered_devices);
+
+  auto tracked_devices = GetLocalDevices();
+  tracked_devices.emplace_back(spmd_device_str);
+  operation_manager_ = std::move(OperationManager(std::move(tracked_devices)));
+}
+
+IfrtComputationClient::~IfrtComputationClient() {
+  // In the GPU case, the PjRtClient depends on the DistributedRuntimeClient
+  // tracked in XlaCoordinator, so the PjRtClient must be destroyed first.
+  client_ = nullptr;
+  coordinator_ = nullptr;
+}
+
+bool IfrtComputationClient::CoordinatorInitialized() const {
+  return coordinator_ != nullptr;
+}
+
+void IfrtComputationClient::InitializeCoordinator(int global_rank,
+                                                  int world_size,
+                                                  std::string master_addr,
+                                                  std::string port) {
+  XLA_CHECK(coordinator_ == nullptr)
+      << "Can only initialize the XlaCoordinator once.";
+  coordinator_ = std::make_unique<XlaCoordinator>(global_rank, world_size,
+                                                  master_addr, port);
+}
+
+XlaCoordinator& IfrtComputationClient::GetCoordinator() {
+  XLA_CHECK(coordinator_ != nullptr)
+      << "XlaCoordinator has not been initialized";
+  return *coordinator_;
+}
+
+void IfrtComputationClient::IfrtData::Assign(
+    const torch::lazy::BackendData& data) {
+  const IfrtData& ifrt_data = dynamic_cast<const IfrtData&>(data);
+  if (&ifrt_data != this) {
+    buffer = ifrt_data.buffer;
+  }
+}
+
+xla::OpSharding IfrtComputationClient::IfrtData::GetSharding() const {
+  XLA_CHECK(HasSharding()) << "Check HasSharding first";
+  return *sharding_;
+}
+
+ComputationClient::DataPtr IfrtComputationClient::CreateDataPlaceholder(
+    std::string device, xla::Shape shape) {
+  return std::make_shared<IfrtData>(device, shape);
+}
+
+std::vector<ComputationClient::DataPtr> IfrtComputationClient::GetDataShards(
+    ComputationClient::DataPtr data) {
+  tsl::profiler::TraceMe activity("IfrtComputationClient::GetDataShards",
+                                  tsl::profiler::TraceMeLevel::kInfo);
+  std::vector<ComputationClient::DataPtr> shards;
+  if (data->HasSharding()) {
+    auto ifrt_data = std::dynamic_pointer_cast<IfrtData>(data);
+    std::vector<tsl::RCReference<xla::ifrt::Array>> arrays =
+        ifrt_data->buffer
+            ->DisassembleIntoSingleDeviceArrays(
+                xla::ifrt::ArrayCopySemantics::kAlwaysCopy)
```

Comment:
Why do we need to copy here?

Reply:
I'm going for safety over performance with this PR. We'll have to go back over this with a careful eye for Array ownership to get the best performance. PJRT let us ignore ownership because aliasing was functionally impossible, but it's very easy to alias underlying data in IFRT.